### PR TITLE
Use absolute_path instead of fs::canonicalize

### PR DESCRIPTION
**Pull request by topecongiro**
This PR is a successor of #3590.

### CODE REVIEW

File: rustfmt-core/rustfmt-lib/tests/source/mods-relative-paths/mod_b.rs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+fn
```
Comment (calebcartwright) [inline rustfmt-core/rustfmt-lib/tests/source/mods-relative-paths/mod_b.rs:1]: Should the `mod_b` file be added to the [skip list](https://github.com/rust-lang/rustfmt/blob/a9d7cd572cc07dccab703c8ce5df4ecbd2f085d4/rustfmt-core/rustfmt-lib/src/test/mod.rs#L23), along with a `recursive` added to the `lib.rs` files?

Comment (topecongiro) [reply rustfmt-core/rustfmt-lib/tests/source/mods-relative-paths/mod_b.rs:1]: Right, good catch. Thank you!

Comment (topecongiro) [reply rustfmt-core/rustfmt-lib/tests/source/mods-relative-paths/mod_b.rs:1]: Fixed in https://github.com/rust-lang/rustfmt/pull/4131/commits/b873aea85d5b66bfb82a10879d73767349333ccd.

**Comment by topecongiro** [comment]
I am still not sure this will work in all cases, but I will merge this for now.

**Comment by rivy** [comment]
@topecongiro , this will likely fail for deep paths even on later version Windows 10 installations. The removal of the MAX_PATH (~260 characters) limitation is *opt-in* and has to have a registry value set (see <https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file#enable-long-paths-in-windows-10-version-1607-and-later>). That's an impossible bar to expect from any significant user population. It *can't* really be counted on at all.

However, there is an available solution... [*extended-paths*](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file#maximum-path-length-limitation). *Extended-paths* (marked by a leading `\\?\` prefix) can be used with versions of the windows API back to at least Windows 7. If the path length is greater than (or even close to) MAX_PATH, it should be converted to an *extended-path* by converting it to an absolute path, converting all forward slashes to backslashes, and prefixing it with `\\?\`.

I [referenced the needed solution](https://github.com/rust-lang/rustfmt/pull/3324#discussion_r283767781) for this in the original issue.

This type of path construction can even be used to access windows wierdo reserved paths like `NUL` and `CON`.

Please see the [`dunce` crate](https://crates.io/crates/dunce) which does a very good job in navigating these weird MS path waters.

**Comment by topecongiro** [comment]
@rivy Hmm, I will take a look again, but the `dance` crate looks promising. We might as well use it. Thank you for your information

**Pull request merged**